### ISSUE TITLE
Second level dependencies missing on for local file packages on >= 6.3.1

### ISSUE THREAD

**Issue by lindycoder** [jazzband/pip-tools#1505]
Hi, thank you for the great tool!

There seems to be a regression with 6.3.1 (and 6.4.0).  Dependencies of dependencies don't make it to the final result when using local packages.

Below is script to reproduce, problem seems to be in pip-tools as any recent tested versions of pip (21.3 and 21.2.4) produce them same problem.

Note : Tested with absolute files to remove the problematic relative files handling issues.

#### Environment Versions

1. OS Type: Reproduced in OSX and a `python:3.8-buster` docker image
1. Python version: `Python 3.8.6`
1. pip version: `pip 21.3 from /private/tmp/pip-tools/venv/lib/python3.8/site-packages/pip (python 3.8)`
1. pip-tools version: `pip-compile, version 6.4.0` && `pip-compile, version 6.3.1`

#### Steps to replicate

```
#!/usr/bin/env bash -x

mkdir /tmp/pip-tools && cd /tmp/pip-tools

mkdir -p project-a
cat > project-a/setup.py <<EOF
#!/usr/bin/env python
from setuptools import setup
setup(
    name="project-a",
    version="0.1",
    install_requires=[
      "pip-tools",
    ]
)
EOF

mkdir -p project-b
cat > project-b/setup.py <<EOF
#!/usr/bin/env python
from setuptools import setup
setup(
    name="project-b",
    version="0.1",
    install_requires=[
      "project-a",
    ]
)
EOF

cat > requirements.in <<EOF
file:///tmp/pip-tools/project-a#egg=project-a
file:///tmp/pip-tools/project-b#egg=project-b
EOF

virtualenv venv
venv/bin/pip install -U "pip==21.3"

venv/bin/pip install -U "pip-tools==6.3.0"
venv/bin/pip-compile -v requirements.in  # Has all reqs

venv/bin/pip install -U "pip-tools==6.3.1"
venv/bin/pip-compile -v requirements.in  # Has no sub reqs

venv/bin/pip install -U "pip-tools==6.4.0"
venv/bin/pip-compile -v requirements.in  # Has no sub reqs
```

#### Expected result
```
# This file is autogenerated by pip-compile with python 3.8
# To update, run:
#
#    pip-compile requirements.in
#
click==8.0.3
    # via pip-tools
pep517==0.11.0
    # via pip-tools
pip-tools==6.4.0
    # via project-a
project-a @ file:///tmp/pip-tools/project-a
    # via
    #   -r requirements.in
    #   project-b
project-b @ file:///tmp/pip-tools/project-b
    # via -r requirements.in
tomli==1.2.1
    # via pep517
wheel==0.37.0
    # via pip-tools

# The following packages are considered to be unsafe in a requirements file:
# pip
# setuptools
```

When it works, verbose log shows round 2 as 
```
New dependencies found in this round:
  adding ('click', '>=7', [])
  adding ('pep517', '', [])
  adding ('pip', '>=21.2', [])
  adding ('setuptools', '', [])
  adding ('wheel', '', [])
Removed dependencies in this round:
------------------------------------------------------------
Result of round 2: not stable
```

#### Actual result
```
# This file is autogenerated by pip-compile with python 3.8
# To update, run:
#
#    pip-compile requirements.in
#
project-a @ file:///tmp/pip-tools/project-a
    # via
    #   -r requirements.in
    #   project-b
project-b @ file:///tmp/pip-tools/project-b
    # via -r requirements.in
```

When it doesn't work, verbose log shows round 2 as 
```
New dependencies found in this round:
  adding ('click', '>=7', [])
  adding ('pep517', '', [])
  adding ('pip', '>=21.2', [])
  adding ('setuptools', '', [])
  adding ('wheel', '', [])
Removed dependencies in this round:
  removing ('pip-tools', '', [])
------------------------------------------------------------
Result of round 2: not stable
```

**Comment by AndydeCleyre** [comment]
Since you've narrowed this down to the `6.3.1` release, I'm going to invite @richafrank to please have a look and a think.

**Comment by richafrank** [comment]
One concern I have is whether that new potential case is caused by the new fix or exists already. If it's real and new, then we'd have to consider which case we prefer breaking until we have a full solution. I'll look into a minimum test case to see if it's real and new.

**Comment by stephan-hof** [comment]
Hi,

unfortunately I'm also affected by this. See here a [test](https://gist.github.com/stephan-hof/e08fcdf582e47dc4fbc91192ed73ce96) which illustrates my situation.

All tests are done with `python3.9` only.

|pip-tools|pip|test result|
|-|-|-|
|HEAD detached at 6.3.0|20.3.4|success|
|HEAD detached at 6.3.1|20.3.4|failed|
|current master (43e532c)|21.2.4|failed|

The observation I made was:
This lines gets called with `ireq` being a new instance. Since `ireq` has no custom hashing, python uses `id(ireq)` as a key. Which is different for each instance, meaning it never hits the cache. https://github.com/jazzband/pip-tools/blob/43e532c2c36d704e8ad90b52c724e525bebfbb9a/piptools/repositories/pypi.py#L222

Since it is not in the cache `self.resolve_reqs` gets called, which ultimately goes into:
https://github.com/pypa/pip/blob/f0f67af32c8f58031d6c239d46899fb824b94262/src/pip/_internal/resolution/legacy/resolver.py#L379

For some reasons (sorry I'm not deep in the pip-tools internals) the `ireq.prepared` is `True`, leading this function to return `[]`. Which means no dependencies, causing the missing dependencies.

Haven't thought of a fix. An idea would be to find a way to make ireq hashable/comparable to make the lookup in the cache working.

FYI: This is `self._dependencies_cache.keys()` of `PyPIRepository` after the test.
See the multiple entries for the test packages.
```
[
<InstallRequirement object: this-will-never-exist-on-pypi-hopefully==0.0.0 from file:///<TMPDIR>/l1 (from -r /../req.in (line 2)) editable=True>,
<InstallRequirement object: depend-on-this-will-never-exist-on-pypi-hopefully==0.0.0 from file:///<TMPDIR>/l2 (from -r /<TMPDIR>/req.in (line 3)) editable=True>,
<InstallRequirement object: this-will-never-exist-on-pypi-hopefully==0.0.0 from file:///<TMPDIR>/l1 (from -r /<TMPDIR>/req.in (line 2)) editable=True>,
<InstallRequirement object: depend-on-this-will-never-exist-on-pypi-hopefully==0.0.0 from file:///<TMPDIR>/l2 (from -r /<TMPDIR>/req.in (line 3)) editable=True>,
<InstallRequirement object: Acquisition==4.10 from <URL> (from this-will-never-exist-on-pypi-hopefully==0.0.0->-r /<TMPDIR>/req.in (line 2)) editable=False>,
<InstallRequirement object: this-will-never-exist-on-pypi-hopefully==0.0.0 from file:///<TMPDIR>/l1 (from -r /<TMPDIR>/req.in (line 2)) editable=True>,
<InstallRequirement object: ExtensionClass==4.6 from <URL> (from Acquisition==4.10->this-will-never-exist-on-pypi-hopefully==0.0.0->-r /<TMPDIR>/req.in (line 2)) editable=False>,
<InstallRequirement object: this-will-never-exist-on-pypi-hopefully==0.0.0 from file:///<TMPDIR>/l1 (from -r /<TMPDIR>/req.in (line 2)) editable=True>,
<InstallRequirement object: zope.interface==5.4.0 from <URL (from Acquisition==4.10->this-will-never-exist-on-pypi-hopefully==0.0.0->-r /<TMPDIR>/req.in (line 2)) editable=False>,
<InstallRequirement object: this-will-never-exist-on-pypi-hopefully==0.0.0 from file:///<TMPDIR>/l1 (from -r /<TMPDIR>/req.in (line 2)) editable=True>,
<InstallRequirement object: setuptools==60.10.0 from <URL> (from zope.interface==5.4.0->Acquisition==4.10->this-will-never-exist-on-pypi-hopefully==0.0.0->-r /<TMPDIR>/req.in (line 2)) editable=False>,
<InstallRequirement object: this-will-never-exist-on-pypi-hopefully==0.0.0 from file:///<TMPDIR>/l1 (from -r /X/req.in (line 2)) editable=True>
]

```

**Comment by AndydeCleyre** [comment]
@stephan-hof Thanks! Can you check your case against #1519?

**Comment by stephan-hof** [comment]
Hi,

thanks for the quick reply. I can confirm that #1519 fixes the test case. Tested it also for my real case with **success**.

Just to let you know: The `_dependencies_cache` in `PyPIRepository` has still multiple entries for the packages.

**Comment by atugushev** [comment]
Does #1539 fix the issue?

**Comment by mmerickel** [comment]
@atugushev  https://github.com/jazzband/pip-tools/pull/1519#issuecomment-1049504651 When I created this comment I had also tested #1539 and it did not fix the issue for me.

**Comment by mtesch-um** [comment]
I'm having the same problem - works ok w/ `pip==20.3.4 pip-tools==6.3.0` but not with latest `6.6.0`.  What happens in the second round is that the requirement gets inexplicably removed in the second round.

What's really strange, and I haven't been able to track down, is that if I put the requirement in one remote repo/package (via setup.py install_requires) it works, but if i put it in a nearly identical setup.py in a different repo/package w/ the same mechanism it does not work.

**Comment by mmerickel** [comment]
What needs to happen to get some movement on this? #1519 has had a fix for this for quite some time now and it's a clear regression in pip-tools behavior between <=6.3.0 and 6.3.1+. We're currently just using 6.3.0 and having to manually comment out the stream assertion error in pypi.py (lol). Our structure is a constraints file some ordered .in files:

```
### base.txt
sqlalchemy < 2.0

### src.in
-c base.txt
-e file:src/our-core-lib#egg=our-core-lib
-e file:src/our-web-app#egg=our-web-app
```

Here is our src.txt in 6.3.0:

```
### src.txt IN 6.3.0
-e file:src/our-core-lib#egg=our-core-lib
  # via
  #    -r src.in
  #    our-web-app
-e file:src/our-web-app#egg=our-web-app
  # via -r src.in
alembic==1.7.6
  # via our-core-lib
sqlalchemy==1.4.32
  # via
  #    alembic
  #    our-core-lib
  #    our-web-app
tabulate==0.8.9
  # via our-core-lib
```

And here is how it changes in 6.3.1+:

```
### src.txt IN 6.3.1 (and 6.6.0)
-e file:src/our-core-lib#egg=our-core-lib
  # via
  #    -r src.in
  #    our-web-app
-e file:src/our-web-app#egg=our-web-app
  # via -r src.in
```

Almost all actual dependencies are missing from the rendered src.txt defeating the entire purpose of pip-tools.

**Comment by atzannes** [comment]
I'm also in an increasing amount of pain because of having to be pinned to an old pip-tools and pip. I hope the fix can get merged and released as soon as possible

**Comment by atugushev** [comment]
The fix is released as part of `pip-tools-6.7.0`.

**Comment by mmerickel** [comment]
Woo, thank you so much for everyone's work - I can confirm at least that 6.8.0 is working for me!
